File: src/main/java/org/janelia/saalfeldlab/paintera/config/ScreenScalesConfig.java
package org.janelia.saalfeldlab.paintera.config;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;

import java.util.Arrays;

public class ScreenScalesConfig {

	private static final double[] DEFAULT_SCREEN_SCALES = new double[] {1.0, 0.5, 0.25, 0.125, 0.0625};

	public static class ScreenScales {

		private final double[] scales;

		public ScreenScales(final ScreenScales scales)
		{
			this(scales.scales.clone());
		}

		public ScreenScales(double... scales) {
			this.scales = scales;
		}

		public double[] getScalesCopy()
		{
			return this.scales.clone();
		}

		@Override
		public int hashCode()
		{
			return Arrays.hashCode(scales);
		}

		@Override
		public boolean equals(Object other)
		{
			return other != null
					&& other instanceof ScreenScales
					&& Arrays.equals(((ScreenScales) other).scales, this.scales);
		}

		@Override
		public String toString()
		{
			return String.format("{ScreenScales: %s}", Arrays.toString(scales));
		}
	}

	private final ObjectProperty<ScreenScales> screenScales = new SimpleObjectProperty<>();

	public ScreenScalesConfig()
	{
		this(DEFAULT_SCREEN_SCALES.clone());
	}

	public ScreenScalesConfig(final double... initialScales)
	{
		this.screenScales.set(new ScreenScales(initialScales.clone()));
	}

	public ObjectProperty<ScreenScales> screenScalesProperty()
	{
		return this.screenScales;
	}

	public void set(final ScreenScalesConfig that)
	{
		this.screenScales.set(that.screenScales.get());
	}

	public static double[] defaultScreenScalesCopy()
	{
		return DEFAULT_SCREEN_SCALES.clone();
	}

	@Override
	public String toString()
	{
		return String.format(
				"{ScreenScalesConfig: %s}",
				this.screenScales.get() == null ? null : this.screenScales.get().toString()
		);
	}

}

File: src/main/java/org/janelia/saalfeldlab/paintera/config/ScreenScalesConfig.kt
package org.janelia.saalfeldlab.paintera.config

import javafx.beans.property.ObjectProperty
import javafx.beans.property.SimpleObjectProperty
import java.util.*

class ScreenScalesConfig @JvmOverloads constructor(vararg initialScales: Double = DEFAULT_SCREEN_SCALES.clone()) {

    private val screenScales = SimpleObjectProperty<ScreenScales>()

    class ScreenScales(vararg scales: Double) {

        private val scales: DoubleArray

        val scalesCopy: DoubleArray
            get() = this.scales.clone()

        constructor(scales: ScreenScales) : this(*scales.scales.clone()) {}

        init {
            this.scales = scales
        }

        override fun hashCode(): Int {
            return Arrays.hashCode(scales)
        }

        override fun equals(other: Any?): Boolean {
            return other is ScreenScales && Arrays.equals(other.scales, this.scales)
        }

        override fun toString(): String {
            return String.format("{ScreenScales: %s}", Arrays.toString(scales))
        }
    }

    init {
        this.screenScales.set(ScreenScales(*initialScales.clone()))
    }

    fun screenScalesProperty(): ObjectProperty<ScreenScales> {
        return this.screenScales
    }

    fun set(that: ScreenScalesConfig) {
        this.screenScales.set(that.screenScales.get())
    }

    override fun toString(): String {
        return String.format(
                "{ScreenScalesConfig: %s}",
                if (this.screenScales.get() == null) null else this.screenScales.get().toString()
        )
    }

    companion object {

        private val DEFAULT_SCREEN_SCALES = doubleArrayOf(1.0, 0.5, 0.25, 0.125, 0.0625)

        @JvmStatic
        fun defaultScreenScalesCopy(): DoubleArray {
            return DEFAULT_SCREEN_SCALES.clone()
        }
    }

}
